Python/optimizer.c

```diff
@@ -935,3 +963,209 @@ PyUnstable_Optimizer_NewUOpOptimizer(void)
+ *
+ * The above analysis assumes perfect hash functions,
+ * but those don't exist, so the real false positive
+ * rates may be worse.
+ */
+
+#define K 6
+
+/* TO DO -- Use more modern hash functions with better distribution of bits */
+static uint32_t
+address_to_hash(void *ptr, uint32_t seed, uint32_t multiplier) {
+    assert(ptr != NULL);
+    uint32_t uhash = seed;
+    uintptr_t addr = (uintptr_t)ptr;
+    for (int i = 0; i < SIZEOF_VOID_P; i++) {
+        uhash ^= addr & 255;
+        uhash *= multiplier;
+        addr >>= 8;
+    }
+    return uhash;
+}
+
+static const uint32_t multipliers[2] = {
+    _PyHASH_MULTIPLIER,
+    110351524,
+};
+
+static const uint32_t seeds[2] = {
+    20221211,
+    2147483647,
+};
+
+void
+_Py_BloomFilter_Init(_PyBloomFilter *bloom)
+{
+    for (int i = 0; i < BLOOM_FILTER_WORDS; i++) {
+        bloom->bits[i] = 0;
+    }
+}
+
+void
+_Py_BloomFilter_Add(_PyBloomFilter *bloom, void *ptr)
+{
+    /* Set k (6) bits */
+    for (int i = 0; i < 2; i++) {
+        uint32_t hash = address_to_hash(ptr, seeds[i], multipliers[i]);
+        for (int j = 0; j < (K/2); j++) {
+            uint8_t bits = hash & 255;
+            bloom->bits[bits >> 5] |= (1 << (bits&31));
+            hash >>= 8;
+        }
+    }
+}
+
+static bool
+bloom_filter_may_contain(_PyBloomFilter *bloom, _PyBloomFilter *hashes)
+{
+    for (int i = 0; i < BLOOM_FILTER_WORDS; i++) {
+        if ((bloom->bits[i] & hashes->bits[i]) != hashes->bits[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+static void
+link_executor(_PyExecutorObject *executor)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    _PyExecutorLinkListNode *links = &executor->vm_data.links;
+    _PyExecutorObject *head = interp->executor_list_head;
+    if (head == NULL) {
+        interp->executor_list_head = executor;
+        links->previous = NULL;
+        links->next = NULL;
+    }
+    else {
+        _PyExecutorObject *next = head->vm_data.links.next;
+        links->previous = head;
+        links->next = next;
+        if (next != NULL) {
+            next->vm_data.links.previous = executor;
+        }
+        head->vm_data.links.next = executor;
+    }
+    executor->vm_data.linked = true;
+}
+
+static void
+unlink_executor(_PyExecutorObject *executor)
+{
+    if (!executor->vm_data.linked) {
+        return;
+    }
+    _PyExecutorLinkListNode *links = &executor->vm_data.links;
+    _PyExecutorObject *next = links->next;
+    _PyExecutorObject *prev = links->previous;
+    if (next != NULL) {
+        next->vm_data.links.previous = prev;
+    }
+    else if (prev == NULL) {
+        // Both are NULL, so this must be the list head.
+        PyInterpreterState *interp = PyInterpreterState_Get();
+        assert(interp->executor_list_head == executor);
+        interp->executor_list_head = executor->vm_data.links.next;
+    }
+    if (prev != NULL) {
+        prev->vm_data.links.next = next;
+    }
```

Comment:
I think this should fix it?

```suggestion
    if (prev != NULL) {
        prev->vm_data.links.next = next;
    }
    else {
        // Prev is NULL, so this must be the list head.
        PyInterpreterState *interp = PyInterpreterState_Get();
        assert(interp->executor_list_head == executor);
        interp->executor_list_head = next;
    }
```

Reply:
This looks correct, thanks.